python/lsst/daf/butler/core/logging.py

```diff
@@ -21,11 +21,51 @@
 
 from __future__ import annotations
 
-__all__ = ("VERBOSE",)
+__all__ = ("VERBOSE", "ButlerMDC")
 
 import logging
 
 VERBOSE = (logging.INFO + logging.DEBUG) // 2
 """Verbose log level"""
 
 logging.addLevelName(VERBOSE, "VERBOSE")
+
+
+class MDCDict(dict):
+    """Dictionary for MDC data.
+
+    This is internal class used for better formatting of MDC in Python logging
+    output. It behaves like `defaultdict(str)` but overrides ``__str__`` and
+    ``__repr__`` method to produce output better suited for logging records.
+    """
+
+    def __getitem__(self, name: str) -> str:
+        """Return value for a given key or empty string for missing key.
+        """
+        return self.get(name, "")
+
+    def __str__(self) -> str:
+        """Return string representation, strings are interpolated without
+        quotes.
+        """
+        items = (f"{k}={self[k]}" for k in sorted(self))
+        return "{" + ", ".join(items) + "}"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+class ButlerMDC:
+    """Handle setting and unsetting of global MDC records."""
+
+    _MDC = MDCDict()
+
+    @classmethod
+    def MDC(cls, key: str, value: str) -> str:
+        old_value = cls._MDC[key]
+        cls._MDC[key] = value
+        return old_value
+
+    @classmethod
+    def MDCRemove(cls, key: str) -> None:
+        del cls._MDC[key]
```

Comment:
Maybe add a protection (`cls._MDC.pop(key, None)`) just in case someone tries to delete it twice. It is probably an error but may not be worth crashing the whole application for that?